fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ScalarApply2AnalyticRule.java

```diff
@@ -0,0 +1,935 @@
+            // E.g. t0.v1 = t1.v4 in the above case
+            List<ScalarOperator> subqueryCorrelationConjuncts =
+                    Utils.extractConjuncts(applyOp.getCorrelationConjuncts());
+            {
+                Iterator<ScalarOperator> outerIt = outerConjuncts.iterator();
+                Iterator<ScalarOperator> subIt = subqueryCorrelationConjuncts.iterator();
+                while (outerIt.hasNext()) {
+                    ScalarOperator outerConjunct = outerIt.next();
+
+                    while (subIt.hasNext()) {
+                        ScalarOperator correlationConjunct = subIt.next();
+                        if (PredicateComparator.isIdentical(correlationConjunct, outerConjunct,
+                                context.getColumnRefFactory())) {
+                            outerCorrelatedConjuncts.add(outerConjunct);
+                            outerIt.remove();
+                            subIt.remove();
+                            break;
+                        }
+                    }
+                }
+            }
+            if (outerCorrelatedConjuncts.isEmpty() || !subqueryCorrelationConjuncts.isEmpty()) {
+                return false;
+            }
+
+            // Second, remove the conjunct that contains the subquery
+            // E.g. t0.v3 < (<subquery>) in the above case
+            {
+                Iterator<ScalarOperator> outerIt = outerConjuncts.iterator();
+                while (outerIt.hasNext()) {
+                    ScalarOperator outerConjunct = outerIt.next();
+                    if (Utils.collect(outerConjunct, ColumnRefOperator.class).stream()
+                            .anyMatch(columnRefOperator -> Objects.equals(applyOp.getOutput(), columnRefOperator))) {
+                        outerSubqueryConjunct = outerConjunct;
+                        outerIt.remove();
+                        break;
+                    }
+                }
+            }
+            if (outerSubqueryConjunct == null) {
+                return false;
+            }
+
+            // Third, remove all the conjuncts which only related to the correlated outer table from the outerConjuncts
+            // E.g. t0.v2 < 5 in the above case
+            {
+                Iterator<ScalarOperator> outerIt = outerConjuncts.iterator();
+                while (outerIt.hasNext()) {
+                    ScalarOperator outerConjunct = outerIt.next();
+                    List<ColumnRefOperator> colRefs = Utils.collect(outerConjunct, ColumnRefOperator.class);
+                    boolean relatedToOtherTable = false;
+                    for (ColumnRefOperator colRef : colRefs) {
+                        Column column = context.getColumnRefFactory().getColumn(colRef);
+                        if (column == null) {
+                            continue;
+                        }
+                        if (!Objects.equals(column, correlatedOuterTable.getColumn(column.getName()))) {
```

Comment:
why not directly call column.equals?

Reply:
Both ok here because column is never null, but I prever Objects.equals which is more generic.